godot-ffi/src/binding/multi_threaded.rs

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Thread safe binding storage.
+//!
+//! This can be used from different threads without issue, as late initialization must be synchronized.
+//!
+//! The user of these structs and functions must still ensure that multi-threaded usage of the various pointers is safe.
+
+use std::sync::OnceLock;
+
+use super::GodotBindingRaw;
+use crate::UnsafeOnceCell;
+
+type GodotBinding = GodotBindingRaw<GdextConfig>;
+
+pub(super) struct BindingStorage {
+    binding: UnsafeOnceCell<GodotBinding>,
+}
+
+impl BindingStorage {
+    /// Get the static binding storage.
+    #[inline(always)]
+    fn storage() -> &'static Self {
+        static BINDING: BindingStorage = BindingStorage {
+            binding: UnsafeOnceCell::new(),
+        };
+        &BINDING
+    }
+
+    /// Initialize the binding storage, this must be called before any other public functions.
+    ///
+    /// # Safety
+    ///
+    /// - Must only be called once.
+    /// - Must not be called concurrently with [`get_binding_unchecked`](BindingStorage::get_binding_unchecked).
+    pub unsafe fn initialize(binding: GodotBinding) -> Result<(), ()> {
+        let storage = Self::storage();
+
+        // SAFETY: `initialize` is only called once, and is not called concurrently with `get_binding_unchecked`, which is the
+        // only place where other methods are called on the binding.
+        unsafe { storage.binding.set(binding) }
+
+        Ok(())
+    }
```

Comment:
i guess since the performance of the initialization function isn't too important, then i could just check that the storage isn't initialized before initializing it. it shouldn't ever fail but we also only call the function once during the entire lifetime of the library.